hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/JoinMessage.java

```diff
@@ -20,16 +20,23 @@
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.version.Version;
 
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 
+// Used as request and response in join protocol
 public class JoinMessage implements IdentifiedDataSerializable {
 
     protected byte packetVersion;
     protected int buildNumber;
+    /**
+     * this is populated with the codebase version of the node trying to join the cluster
+     * (ie {@link com.hazelcast.instance.Node#getVersion()}).
+     */
+    protected Version version;
     protected Address address;
```

Comment:
Could we describe somewhere the flow of JoinMessages and joinRequests, and responses, also including what happenes with the version? It will be easier to understand it for people who know nothing about it.

It's still a bit confusing to me what the difference is between `JoinMessage` and `JoinRequest`